### ISSUE TITLE
Large directories are working slowly (`cd`, `ls -1`)

### ISSUE THREAD

**Issue by alexey-milovidov** [yandex-cloud/geesefs#28]
I have a directory with 100 000 files.
Simply doing `cd` to it from `bash` takes minutes.

Opening a directory in `mc` may take tens of minutes.

**Comment by alexey-milovidov** [comment]
I underestimated it.
Looks like it did not open a directory in 30 minutes.

**Comment by vitalif** [comment]
Hi
I suspect you're doing it under macos?) Try to use Linux if so.
As far as I remember there's a problem that MacOS tries to read xattrs of all files when listing a directory.
With Yandex Object Storage it isn't a problem because we have an extended listing format which includes user metadata.
But in the previous bug you said you were using AWS, and AWS doesn't have user metadata in listings. Thus if the OS tries to read xattrs geesefs starts to make HEAD requests for each file in the directory.
If you think my guess is wrong - try to mount geesefs with `--debug_fuse --debug_s3 --log-file /path/to/log.txt`, reproduce the slow listing and attach the resulting debug log here.

**Comment by vitalif** [comment]
If it's the case with xattrs and macos, I can add an option to disable xattrs as a workaround :-)

**Comment by alexey-milovidov** [comment]
I'm doing it in Linux.

Still experimenting. The last try to call `ls -1` did not finish in 45 minutes.
Maybe it is fetching some metadata for every file sequentially.
Maybe it is doing readdir with small buffer?

So, the solution can be in prefetching the directory listing and files metadata asynchronously?

**Comment by alexey-milovidov** [comment]
```
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ time ls -1 | pv -l | wc -l
^C00  0:48:01 [0.00 /s] [<=>                                                                                                                                                                               ]
^C^C^C

real    48m3.111s
user    0m0.491s                                                                                                                                                                                            
sys     0m1.169s                                                                                                                                                                                            
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ ^C                                                                                                                                                                    
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ time find . | pv -l | wc -l                                                                                                                                           
^C00  0:05:22 [0.00 /s] [<=>                                                                                                                                                                               ]
^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C                                                                                
                                                                                                                                                                                                            
real    5m27.354s                                                                                                                                                                                           
user    0m0.073s                                                                                                                                                                                            
sys     0m0.117s                                                                                                                                                                                            
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ ^C                                                                                                                                                                    
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ ^C                                                                                                                                                                    
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ ^C                                                                                                                                                                    
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ ^C                                                                                                                                                                    
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ ^C                                                                                                                                                                    
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ ^C                                                                                                                                                                    
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ ^C
ubuntu@ip-172-31-32-168:~/s3fuse/zip$ strace -T -f find . 
execve("/usr/bin/find", ["find", "."], 0x7ffc3ab6e138 /* 25 vars */) = 0 <0.000214>
brk(NULL)                               = 0x55906516f000 <0.000008>
arch_prctl(0x3001 /* ARCH_??? */, 0x7ffc90e2c1f0) = -1 EINVAL (Invalid argument) <0.000052>
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory) <0.000015>
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3 <0.000032>
fstat(3, {st_mode=S_IFREG|0644, st_size=31765, ...}) = 0 <0.000008>
mmap(NULL, 31765, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f638e79d000 <0.000010>
close(3)                                = 0 <0.000007>
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libselinux.so.1", O_RDONLY|O_CLOEXEC) = 3 <0.000012>
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0@p\0\0\0\0\0\0"..., 832) = 832 <0.000009>
fstat(3, {st_mode=S_IFREG|0644, st_size=163200, ...}) = 0 <0.000008>
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f638e79b000 <0.000010>
mmap(NULL, 174600, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f638e770000 <0.000011>
mprotect(0x7f638e776000, 135168, PROT_NONE) = 0 <0.000013>
mmap(0x7f638e776000, 102400, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x6000) = 0x7f638e776000 <0.000014>
mmap(0x7f638e78f000, 28672, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1f000) = 0x7f638e78f000 <0.000038>
mmap(0x7f638e797000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x26000) = 0x7f638e797000 <0.000013>
mmap(0x7f638e799000, 6664, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f638e799000 <0.000011>
close(3)                                = 0 <0.000008>
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libm.so.6", O_RDONLY|O_CLOEXEC) = 3 <0.000016>
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\300\323\0\0\0\0\0\0"..., 832) = 832 <0.000008>
fstat(3, {st_mode=S_IFREG|0644, st_size=1369352, ...}) = 0 <0.000007>
mmap(NULL, 1368336, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f638e621000 <0.000010>
mmap(0x7f638e62e000, 684032, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0xd000) = 0x7f638e62e000 <0.000014>
mmap(0x7f638e6d5000, 626688, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0xb4000) = 0x7f638e6d5000 <0.000012>
mmap(0x7f638e76e000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x14c000) = 0x7f638e76e000 <0.000011>
close(3)                                = 0 <0.000036>
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3 <0.000014>
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\360A\2\0\0\0\0\0"..., 832) = 832 <0.000008>
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784 <0.000007>
pread64(3, "\4\0\0\0\20\0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0", 32, 848) = 32 <0.000007>
pread64(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\237\333t\347\262\27\320l\223\27*\202C\370T\177"..., 68, 880) = 68 <0.000007>
fstat(3, {st_mode=S_IFREG|0755, st_size=2029560, ...}) = 0 <0.000007>
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784 <0.000007>
pread64(3, "\4\0\0\0\20\0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0", 32, 848) = 32 <0.000008>
pread64(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\237\333t\347\262\27\320l\223\27*\202C\370T\177"..., 68, 880) = 68 <0.000007>
mmap(NULL, 2037344, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f638e42f000 <0.000011>
mmap(0x7f638e451000, 1540096, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x22000) = 0x7f638e451000 <0.000014>
mmap(0x7f638e5c9000, 319488, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x19a000) = 0x7f638e5c9000 <0.000011>
mmap(0x7f638e617000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1e7000) = 0x7f638e617000 <0.000012>
mmap(0x7f638e61d000, 13920, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f638e61d000 <0.000010>
close(3)                                = 0 <0.000006>
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libpcre2-8.so.0", O_RDONLY|O_CLOEXEC) = 3 <0.000014>
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\340\"\0\0\0\0\0\0"..., 832) = 832 <0.000008>
fstat(3, {st_mode=S_IFREG|0644, st_size=584392, ...}) = 0 <0.000008>
mmap(NULL, 586536, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f638e39f000 <0.000011>
mmap(0x7f638e3a1000, 409600, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x2000) = 0x7f638e3a1000 <0.000017>
mmap(0x7f638e405000, 163840, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x66000) = 0x7f638e405000 <0.000012>
mmap(0x7f638e42d000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x8d000) = 0x7f638e42d000 <0.000012>
close(3)                                = 0 <0.000038>
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libdl.so.2", O_RDONLY|O_CLOEXEC) = 3 <0.000015>
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0 \22\0\0\0\0\0\0"..., 832) = 832 <0.000009>
fstat(3, {st_mode=S_IFREG|0644, st_size=18816, ...}) = 0 <0.000008>
mmap(NULL, 20752, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f638e399000 <0.000012>
mmap(0x7f638e39a000, 8192, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1000) = 0x7f638e39a000 <0.000016>
mmap(0x7f638e39c000, 4096, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x3000) = 0x7f638e39c000 <0.000012>
mmap(0x7f638e39d000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x3000) = 0x7f638e39d000 <0.000011>
close(3)                                = 0 <0.000008>
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libpthread.so.0", O_RDONLY|O_CLOEXEC) = 3 <0.000014>
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\220q\0\0\0\0\0\0"..., 832) = 832 <0.000007>
pread64(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\360\2300%\360\340\363'\246\332u/\364\377\246u"..., 68, 824) = 68 <0.000008>
fstat(3, {st_mode=S_IFREG|0755, st_size=157224, ...}) = 0 <0.000008>
pread64(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\360\2300%\360\340\363'\246\332u/\364\377\246u"..., 68, 824) = 68 <0.000008>
mmap(NULL, 140408, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f638e376000 <0.000011>
mmap(0x7f638e37c000, 69632, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x6000) = 0x7f638e37c000 <0.000016>
mmap(0x7f638e38d000, 24576, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x17000) = 0x7f638e38d000 <0.000011>
mmap(0x7f638e393000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1c000) = 0x7f638e393000 <0.000013>
mmap(0x7f638e395000, 13432, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f638e395000 <0.000011>
close(3)                                = 0 <0.000007>
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f638e374000 <0.000009>
mmap(NULL, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f638e371000 <0.000053>
arch_prctl(ARCH_SET_FS, 0x7f638e371800) = 0 <0.000008>
mprotect(0x7f638e617000, 16384, PROT_READ) = 0 <0.000014>
mprotect(0x7f638e393000, 4096, PROT_READ) = 0 <0.000010>
mprotect(0x7f638e39d000, 4096, PROT_READ) = 0 <0.000009>
mprotect(0x7f638e42d000, 4096, PROT_READ) = 0 <0.000009>
mprotect(0x7f638e76e000, 4096, PROT_READ) = 0 <0.000009>
mprotect(0x7f638e797000, 4096, PROT_READ) = 0 <0.000009>
mprotect(0x5590633b5000, 8192, PROT_READ) = 0 <0.000011>
mprotect(0x7f638e7d2000, 4096, PROT_READ) = 0 <0.000011>
munmap(0x7f638e79d000, 31765)           = 0 <0.000018>
set_tid_address(0x7f638e371ad0)         = 2826538 <0.000007>
set_robust_list(0x7f638e371ae0, 24)     = 0 <0.000007>
rt_sigaction(SIGRTMIN, {sa_handler=0x7f638e37cbf0, sa_mask=[], sa_flags=SA_RESTORER|SA_SIGINFO, sa_restorer=0x7f638e38a3c0}, NULL, 8) = 0 <0.000008>
rt_sigaction(SIGRT_1, {sa_handler=0x7f638e37cc90, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_SIGINFO, sa_restorer=0x7f638e38a3c0}, NULL, 8) = 0 <0.000007>
rt_sigprocmask(SIG_UNBLOCK, [RTMIN RT_1], NULL, 8) = 0 <0.000007>
prlimit64(0, RLIMIT_STACK, NULL, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0 <0.000060>
statfs("/sys/fs/selinux", 0x7ffc90e2c140) = -1 ENOENT (No such file or directory) <0.000020>
statfs("/selinux", 0x7ffc90e2c140)      = -1 ENOENT (No such file or directory) <0.000009>
brk(NULL)                               = 0x55906516f000 <0.000007>
brk(0x559065190000)                     = 0x559065190000 <0.000010>
openat(AT_FDCWD, "/proc/filesystems", O_RDONLY|O_CLOEXEC) = 3 <0.000016>
fstat(3, {st_mode=S_IFREG|0444, st_size=0, ...}) = 0 <0.000007>
read(3, "nodev\tsysfs\nnodev\ttmpfs\nnodev\tbd"..., 1024) = 384 <0.000026>
read(3, "", 1024)                       = 0 <0.000007>
close(3)                                = 0 <0.000010>
access("/etc/selinux/config", F_OK)     = -1 ENOENT (No such file or directory) <0.000012>
openat(AT_FDCWD, ".", O_RDONLY|O_CLOEXEC) = 3 <0.000233>
uname({sysname="Linux", nodename="ip-172-31-32-168", ...}) = 0 <0.000012>
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0 <0.000016>
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 4 <0.000022>
fstat(4, {st_mode=S_IFREG|0644, st_size=3035952, ...}) = 0 <0.000012>
mmap(NULL, 3035952, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638e08b000 <0.000018>
close(4)                                = 0 <0.000012>
openat(AT_FDCWD, "/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = 4 <0.000021>
fstat(4, {st_mode=S_IFREG|0644, st_size=2996, ...}) = 0 <0.000012>
read(4, "# Locale name alias data base.\n#"..., 4096) = 2996 <0.000014>
read(4, "", 4096)                       = 0 <0.000033>
close(4)                                = 0 <0.000013>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = 4 <0.000019>
fstat(4, {st_mode=S_IFREG|0644, st_size=252, ...}) = 0 <0.000012>
mmap(NULL, 252, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638e7d1000 <0.000014>
close(4)                                = 0 <0.000011>
openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/gconv/gconv-modules.cache", O_RDONLY) = 4 <0.000019>
fstat(4, {st_mode=S_IFREG|0644, st_size=27002, ...}) = 0 <0.000012>
mmap(NULL, 27002, PROT_READ, MAP_SHARED, 4, 0) = 0x7f638e79e000 <0.000014>
close(4)                                = 0 <0.000011>
futex(0x7f638e61c954, FUTEX_WAKE_PRIVATE, 2147483647) = 0 <0.000010>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_MEASUREMENT", O_RDONLY|O_CLOEXEC) = 4 <0.000019>
fstat(4, {st_mode=S_IFREG|0644, st_size=23, ...}) = 0 <0.000012>
mmap(NULL, 23, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638e79d000 <0.000015>
close(4)                                = 0 <0.000011>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_TELEPHONE", O_RDONLY|O_CLOEXEC) = 4 <0.000019>
fstat(4, {st_mode=S_IFREG|0644, st_size=47, ...}) = 0 <0.000011>
mmap(NULL, 47, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638e08a000 <0.000014>
close(4)                                = 0 <0.000011>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_ADDRESS", O_RDONLY|O_CLOEXEC) = 4 <0.000019>
fstat(4, {st_mode=S_IFREG|0644, st_size=131, ...}) = 0 <0.000012>
mmap(NULL, 131, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638e089000 <0.000014>
close(4)                                = 0 <0.000011>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_NAME", O_RDONLY|O_CLOEXEC) = 4 <0.000058>
fstat(4, {st_mode=S_IFREG|0644, st_size=62, ...}) = 0 <0.000061>
mmap(NULL, 62, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638e088000 <0.000012>
close(4)                                = 0 <0.000007>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_PAPER", O_RDONLY|O_CLOEXEC) = 4 <0.000013>
fstat(4, {st_mode=S_IFREG|0644, st_size=34, ...}) = 0 <0.000059>
mmap(NULL, 34, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638e087000 <0.000011>
close(4)                                = 0 <0.000007>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_MESSAGES", O_RDONLY|O_CLOEXEC) = 4 <0.000013>
fstat(4, {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0 <0.000008>
close(4)                                = 0 <0.000009>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_MESSAGES/SYS_LC_MESSAGES", O_RDONLY|O_CLOEXEC) = 4 <0.000011>
fstat(4, {st_mode=S_IFREG|0644, st_size=48, ...}) = 0 <0.000007>
mmap(NULL, 48, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638e086000 <0.000010>
close(4)                                = 0 <0.000006>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_MONETARY", O_RDONLY|O_CLOEXEC) = 4 <0.000013>
fstat(4, {st_mode=S_IFREG|0644, st_size=270, ...}) = 0 <0.000007>
mmap(NULL, 270, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638e085000 <0.000010>
close(4)                                = 0 <0.000007>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_COLLATE", O_RDONLY|O_CLOEXEC) = 4 <0.000012>
fstat(4, {st_mode=S_IFREG|0644, st_size=1518110, ...}) = 0 <0.000007>
mmap(NULL, 1518110, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638df12000 <0.000083>
close(4)                                = 0 <0.000008>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_TIME", O_RDONLY|O_CLOEXEC) = 4 <0.000015>
fstat(4, {st_mode=S_IFREG|0644, st_size=3360, ...}) = 0 <0.000009>
mmap(NULL, 3360, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638df11000 <0.000012>
close(4)                                = 0 <0.000008>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_NUMERIC", O_RDONLY|O_CLOEXEC) = 4 <0.000013>
fstat(4, {st_mode=S_IFREG|0644, st_size=50, ...}) = 0 <0.000007>
mmap(NULL, 50, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638df10000 <0.000010>
close(4)                                = 0 <0.000006>
openat(AT_FDCWD, "/usr/lib/locale/C.UTF-8/LC_CTYPE", O_RDONLY|O_CLOEXEC) = 4 <0.000011>
fstat(4, {st_mode=S_IFREG|0644, st_size=201272, ...}) = 0 <0.000007>
mmap(NULL, 201272, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f638dede000 <0.000011>
close(4)                                = 0 <0.000007>
openat(AT_FDCWD, "/usr/share/locale/C.UTF-8/LC_MESSAGES/findutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000013>
openat(AT_FDCWD, "/usr/share/locale/C.utf8/LC_MESSAGES/findutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000010>
openat(AT_FDCWD, "/usr/share/locale/C/LC_MESSAGES/findutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000010>
openat(AT_FDCWD, "/usr/share/locale-langpack/C.UTF-8/LC_MESSAGES/findutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000009>
openat(AT_FDCWD, "/usr/share/locale-langpack/C.utf8/LC_MESSAGES/findutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000009>
openat(AT_FDCWD, "/usr/share/locale-langpack/C/LC_MESSAGES/findutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000008>
ioctl(1, TCGETS, {B38400 opost isig icanon echo ...}) = 0 <0.000008>
newfstatat(AT_FDCWD, ".", {st_mode=S_IFDIR|0755, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0 <0.000207>
fstat(1, {st_mode=S_IFCHR|0620, st_rdev=makedev(0x88, 0x2), ...}) = 0 <0.000014>
write(1, ".\n", 2.
)                      = 2 <0.000013>
openat(AT_FDCWD, ".", O_RDONLY|O_NOCTTY|O_NONBLOCK|O_NOFOLLOW|O_CLOEXEC|O_DIRECTORY) = 4 <0.000157>
fstat(4, {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0 <0.000012>
fcntl(4, F_GETFL)                       = 0x38800 (flags O_RDONLY|O_NONBLOCK|O_LARGEFILE|O_NOFOLLOW|O_DIRECTORY) <0.000072>
fcntl(4, F_SETFD, FD_CLOEXEC)           = 0 <0.000085>
newfstatat(AT_FDCWD, ".", {st_mode=S_IFDIR|0755, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0 <0.000092>
fstatfs(4, {f_type=FUSE_SUPER_MAGIC, f_bsize=1048576, f_blocks=274877906944, f_bfree=274877906944, f_bavail=274877906944, f_files=1000000000, f_ffree=1000000000, f_fsid={val=[0, 0]}, f_namelen=255, f_frsize=4096, f_flags=ST_VALID|ST_NOSUID|ST_NODEV|ST_RELATIME}) = 0 <0.000270>
fcntl(4, F_DUPFD_CLOEXEC, 3)            = 5 <0.000110>
getdents64(4, /* 79 entries */, 32768)  = 3808 <9.327656>
getdents64(4, /* 79 entries */, 32768)  = 3760 <9.725057>
getdents64(4, /* 81 entries */, 32768)  = 3808 <9.848062>
getdents64(4
```

**Comment by alexey-milovidov** [comment]
So, it is receiving around 3800 directory entries in around 10 seconds (380 entries/sec).
For a directory with 100 000 entries it should take around 5 minutes.

I understand that caching is not possible in this operation...

**Comment by alexey-milovidov** [comment]
geesefs binary eats around 30% of CPU core during this operation - looks normal (it's not a bottleneck) but still quite high.

**Comment by vitalif** [comment]
Can you please collect & attach the debug log
Of course normally it doesn't make requests for separate files and normally it uses 1000 file batches :-)
And it also uses caching, i.e. the directory listing is loaded in memory and served from there.
So I need the log to find the problem :-) at least partial if you don't want to wait for the listing to complete

**Comment by alexey-milovidov** [comment]
Ok. The command still runs after 3 hours. I can wait a few hours more, then will enable debug.

**Comment by vitalif** [comment]
:-)) you don't have to wait for so long, I think ~5 minutes of running listing will be enough to debug

**Comment by alexey-milovidov** [comment]
I cannot unmount now, because I have a process downloading these files. So let's wait while it will download a million more files and then debug...

**Comment by vitalif** [comment]
> I cannot unmount now, because I have a process downloading these files. So let's wait while it will download a million more files and then debug...

You can mount another instance with debug in parallel, try to read the directory, wait for several minutes and then just kill the geesefs process, after that the mountpoint will return "transport endpoint is not connected", but you'll be able to unmount it. Ok, anyway, I'm waiting for the log :-)

**Comment by vitalif** [comment]
? :-)

**Comment by almson** [comment]
It's because the default listing strategy is Yandex-proprietary `ext-v1`. This causes a lot of request failures and retries. Setting `--list-type 2` fixes the issue with AWS S3. I don't think geesefs should try to use ext-v1 unless it detects a yandex endpoint URL, or it's set explicitly.

**Comment by vitalif** [comment]
> It's because the default listing strategy is Yandex-proprietary `ext-v1`. This causes a lot of request failures and retries. Setting `--list-type 2` fixes the issue with AWS S3. I don't think geesefs should try to use ext-v1 unless it detects a yandex endpoint URL, or it's set explicitly.

Oops, sorry, geesefs really had a problem with detection of fallback listing method requirement. I fixed it in master, now it automatically falls back to v1 listings.

**Comment by almson** [comment]
Some experiments to list a 50k object AWS S3 bucket in us-east-1 from NYC.

  - Java SDK (version 1.11.883) using v1 API: **8 s** (~20 Mbps network usage)
  - `aws s3 ls ... | wc` **22 s** (~5 Mbps)
  - geesefs 0.30.9 `ls -l ... | wc` without specifying `list-type`: **more than 15 minutes, didn't wait** (~20 Mbps)
  - geesefs 0.30.9 with `--list-type 1`:
     - **returns only 15k results** after 70s
     - run `ls -l ... | wc` again, takes 3 minutes while pegging 100% CPU and using network sporadically, also returns ~15k results
     - wait for CPU to die down (a minute more)
     - `ls -l ... | wc` returns in **23 s** (5 Mbps) with correct result
  - `--list-type 2` similar to `1`, except at the end it takes **12-15 s**

So default `ext-v1` is much slower, but even `1`/`2` has a serious bug!

Is there a reason it can't be as fast as the Java SDK? (Hint: Java probably uses connection pooling.)

**Comment by almson** [comment]
I decided to play around with FUSE using the `jnr-fuse` java library. Unfortunately, it's only FUSE 2, but I noticed that in FUSE 3 it's possible to set `fuse_fill_dir_flags` during `readdir` to return the file stats hopefully avoiding the overhead of `getattr` that `ls -l` implies. Also there's a `cache_readdir` flag in `fuse_file_info`.

**Comment by almson** [comment]
Do you even test with AWS? Why don't you put a warning at the top, "Only works with Yandex"?

**Comment by vitalif** [comment]
Integration tests are run against an S3 emulator that doesn't support that yandex-specific feature, tests run fine :-). List-type is set explicitly there though, that's why incorrect autodetection bug got unnoticed.

**Comment by alexey-milovidov** [comment]
@almson I use it successfully with AWS.

**Comment by vitalif** [comment]
The list-type detection fix is fixed since 0.31.0, so I'm closing this issue by now.
